### DISCUSSION TITLE
Stream created with `ZipWriter` is not read by `Response`

### DISCUSSION THREAD

**Discussion by sntran** [gildas-lormeau/zip.js#407] [Q&A] [answered]
Using the [Hello World with Streams example](https://github.com/gildas-lormeau/zip.js#hello-world-with-streams), I try to write a server that responds with zip file, but it does not work.

The minimal code for such server is below:

```ts
import { serve } from "https://deno.land/std@0.177.0/http/server.ts";

import {
  ZipWriter,
} from "https://deno.land/x/zipjs@v2.6.79/index.js";

serve(async (req: Request) => {
  // Creates a TransformStream object, the zip content will be written in the
  // `writable` property. The `readable` property is responded to client.
  const { readable, writable } = new TransformStream();
  const zipWriter = new ZipWriter(writable);
  const entry = await zipWriter.add("hello.txt", new Blob(["Hello world!"]).stream());
  console.log(entry);
  await zipWriter.close();

  const headers = new Headers();
  headers.set("Content-Disposition", `attachment; filename="hello.zip"`);

  return new Response(readable, {
    // headers,
  });
});
```

You can also see it in action on Deno Deploy Playground: https://dash.deno.com/playground/zipjs

The `console.log(entry)` is never called. The whole process just hangs.

**Comment by gildas-lormeau** [comment, marked as answer]
This is due to the usage of `await`. Here's an example showing how to make it work: https://dash.deno.com/playground/bitter-kingfisher-23

**Reply by gildas-lormeau** [reply]
The problem is that zip.js cannot write to the writable stream because the readable stream is not consumed. An alternative is to define the `highWaterMark`of the writable stream to the desired size (if known), see https://dash.deno.com/playground/interesting-oyster-43.

**Reply by sntran** [reply]
> This is due to the usage of `await`. Here's an example showing how to make it work: https://dash.deno.com/playground/bitter-kingfisher-23

That was it. I did try not to use `await` on `add`, but forgot about the `close` too. Thanks a lot!

**Reply by sntran** [reply]
> The problem is that zip.js cannot write to the writable stream because the readable stream is not consumed. An alternative is to define the `highWaterMark`of the writable stream to the desired size (if known), see https://dash.deno.com/playground/interesting-oyster-43.

It should not be a problem for my case, since the readable stream will be immediately consumed by the client, but good to know!

**Reply by gildas-lormeau** [reply]
The usage of `await` blocks the Response object to read the `readable`, and thus it prevents zip.js to write in to the `writable`. Defining the `highWaterMark`value  prevents zip.js to be blocked.
